File: backend/chart_maps/news_tab.py
"""📰 News tab on Chart Maps — four served reads, one payload (Ajay 2026-09-24).

Ajay 2026-09-24: "build me a news tab in chartmaps to give me a bullish market
or bearsish market and also pull Macro calendar that has T1 and T2 tier events
in to this tab consider in to news. If bullish or beaish I need to whcih
sectors are bullish or which hotsectors are bearish. In a table."

Four blocks, each read from an engine that already ships, each budgeted on
its own and each failing on its own:

  * verdict   — the Market Gauge's daily AND weekly state (`sepa.market_gauge`),
                its word mapped ONCE in `MARKET_WORD`. Nothing is re-scored.
  * macro     — T1 + T2 rows of the FRED-scheduled calendar, read at the
                calendar's ONE cache key (`macro_calendar.DEFAULT_DAYS`). This
                module never passes a window length to the calendar: a second
                key would evict the gauge's doc and force a cold FRED fetch.
  * sectors   — the 🔥 Hottest board's sector rows (`rotation.hottest`),
                the 🔥/🧊 heat word from `rotation.heat`, the served 📰 day tag
                and the board's own `d1` block, which says whether the day
                column is the live session or the last close.
  * headlines — `news_search.core`, the app's ONE news routine.
  * model_read — 🧠 the local abliterated model's stored bull + bear read of
                the four blocks above (`news_model_read`, 2026-09-24). Served
                from Mongo; the model runs only in a background thread.

NOT A SIGNAL. Nothing here is measured to predict, nothing gates a scan,
pushes a phone, sizes a position or enters a lane. The sector-heat study is
served beside the table with its confidence intervals; the 5-session heat
that decides the word is UNMEASURED.
"""
from __future__ import annotations

import asyncio
import logging
import math
import time
from typing import Optional

import macro_calendar
from news_search import core
from rotation import heat
from rotation import hottest as H
from rotation import tracker as T

from . import news_model_read

log = logging.getLogger("chart_maps.news_tab")

# THE one map from the gauge's state key to the word the tab prints.
MARKET_WORD = {"constructive": "bullish", "caution": "mixed", "risk_off": "bearish"}
UNKNOWN_WORD = "unknown"
AUDIT_TAG = "chart_maps_news"
# news.py's own Google query for the market feed, reused as a keyword through
# the ONE news routine — not a second fetcher.
MARKET_QUERY = "stock market OR Nasdaq OR S&P 500"
# ONE request budget, applied to EVERY leg. A leg past it serves ok=False with
# a reason; its thread keeps running so the module caches (gauge, 6 h
# calendar, rotation doc) still warm for the next refresh.
LEG_BUDGET_SEC = 8.0
TIMED_OUT = "still loading — timed out after {:.0f}s, refresh"
NOTE = ("A read of what the app already serves — the Market Gauge's state, the FRED-scheduled macro calendar, "
        "the rotation grid's sector medians against RSP, and the last 36 hours of market headlines. "
        "Nothing here is a forecast, nothing here is measured to predict, and nothing here gates a scan, "
        "pushes a phone, sizes a position or enters a lane. Not advice.")
SECTOR_STUDY = {
    "measured_on": "rel_21d", "shipped_on": heat.HEAT_KEY, "date": "2026-09-09",
    "arrivals": 50191, "dates": 192, "names": 2243,
    "hot_win_rate_pp": -0.57, "hot_win_rate_ci": [-1.87, 0.71],
    "hot_minus_cold_5d_pp": -2.55, "hot_minus_cold_5d_ci": [-4.48, -0.65],
    "script": "backend/studies/sector_heat_study.py", "doc": "docs/supply_demand/sector_heat.md",
    "note": ("Sector heat describes what already moved; it is not a forecast. MEASURED 2026-09-09 on the "
             "rel_21d definition over 50,191 demand-zone arrivals: a hot sector changed the win rate by "
             "−0.57pp (95% CI −1.87 to +0.71) — nothing. 'A cold sector just sits there' measured INVERTED: "
             "at 5 sessions hot minus cold was −2.55pp (95% CI −4.48 to −0.65), the one interval in the study "
             "clear of zero — cold turned faster. The 🔥/🧊 word on this table is decided on the 5-session leg "
             "(rel_5d), which is UNMEASURED; the 21-day null neither validates nor condemns it. "
             "Script: backend/studies/sector_heat_study.py."),
}
_NO_D1 = "no day block served"


def _num(v) -> Optional[float]:
    """Finite floats only; bools, junk, NaN and inf are None."""
    if isinstance(v, bool):
        return None
    try:
        f = float(v)
    except (TypeError, ValueError):
        return None
    return f if math.isfinite(f) else None


def market_word(state) -> str:
    """The gauge state key → bullish / mixed / bearish. Exact key only; any
    other value (a label, a word, a number, None) is UNKNOWN. Never raises."""
    if not isinstance(state, str):
        return UNKNOWN_WORD
    return MARKET_WORD.get(state, UNKNOWN_WORD)


def leg_word(v) -> str:
    """Sign of a relative-strength leg vs the benchmark. 0 is flat."""
    f = _num(v)
    if f is None:
        return UNKNOWN_WORD
    if f > 0:
        return "bullish"
    if f < 0:
        return "bearish"
    return "flat"


def _gauge_card(g: dict) -> dict:
    state = g.get("state")
    return {"score": g.get("score"), "state": state,
            "state_label": g.get("state_label"), "word": market_word(state)}


def verdict_block(gauge: Optional[dict]) -> dict:
    """The gauge's daily and weekly state, each with its word. PURE."""
    if not isinstance(gauge, dict) or not gauge:
        return {"ok": False, "reason": "market gauge unavailable"}
    daily = _gauge_card(gauge)
    wk = gauge.get("weekly")
    weekly = _gauge_card(wk) if isinstance(wk, dict) and wk else None
    outlook = gauge.get("next_day_outlook") or {}
    if not isinstance(outlook, dict):
        outlook = {}
    return {
        "ok": True,
        "as_of_label": gauge.get("as_of_label"),
        "generated_at_iso": gauge.get("generated_at_iso"),
        "daily": daily,
        "weekly": weekly,
        "agree": None if weekly is None else (daily["state"] == weekly["state"]),
        "drivers": list(gauge.get("drivers") or []),
        "outlook": {"label": outlook.get("label"), "note": outlook.get("note"),
                    "watch": list(outlook.get("watch") or [])},
        "disclaimer": gauge.get("disclaimer"),
    }


def macro_block() -> dict:
    """T1 + T2 macro rows at the calendar's ONE cache key.

    Both calls take the calendar's defaults — never a window argument (C1):
    `imminent_events` reads `get_macro_calendar()` at its default key, so a
    different key here would evict that doc and pay a second cold compute.
    """
    try:
        cal = macro_calendar.get_macro_calendar() or {}
        events = macro_calendar.imminent_events(within_days=macro_calendar.DEFAULT_DAYS,
                                                max_tier=2)
    except Exception as exc:                                   # noqa: BLE001
        log.warning("news_tab: macro calendar unavailable: %s", exc)
        return {"ok": False, "reason": f"macro calendar unavailable ({type(exc).__name__})",
                "days": macro_calendar.DEFAULT_DAYS, "events": []}
    rows = []
    for e in events or []:
        tier = e.get("tier")
        if tier not in (1, 2):
            continue
        rows.append({"date": e.get("date"), "kind": e.get("kind"), "tier": tier,
                     "tier_label": macro_calendar.TIER_LABELS.get(tier),
                     "label": e.get("label"), "days_until": e.get("days_until"),
                     "when_label": e.get("when_label")})
    labels = cal.get("tier_labels") or macro_calendar.TIER_LABELS
    return {
        "ok": True,
        "days": macro_calendar.DEFAULT_DAYS,
        "tier_labels": {str(k): v for k, v in labels.items()},
        "next_tier1": cal.get("next_tier1"),
        "events": rows,
        "disclaimer": cal.get("disclaimer"),
    }


def d1_block(body: dict) -> dict:
    """What the day column is showing, from the board's own `d1` block. PURE.
    `moves` (one float per priced name) is never carried."""
    d1 = (body or {}).get(H.D1_KEY) if isinstance(body, dict) else None
    if not isinstance(d1, dict) or not d1:
        return {"live": False, "basis": None, "as_of": None,
                "reason": _NO_D1, "market_closed": None}
    return {"live": d1.get("live") is True, "basis": d1.get("basis"),
            "as_of": d1.get("as_of"), "reason": d1.get("reason"),
            "market_closed": d1.get("market_closed")}


def _heat_of(sector, index: dict) -> dict:
    r = None
    try:
        r = heat.read("", index, sector=sector) if sector else None
    except Exception as exc:                                   # noqa: BLE001
        log.debug("news_tab: heat read %s failed: %s", sector, exc)
    if not isinstance(r, dict):
        return {"tone": UNKNOWN_WORD, "percentile": None,
                "heat_window": heat.HEAT_WINDOW, "thin": None, "grain": None}
    return {"tone": r.get("tone") or UNKNOWN_WORD, "percentile": r.get("percentile"),
            "heat_window": r.get("heat_window"), "thin": r.get("thin"),
            "grain": r.get("grain")}


def sector_rows(body: dict, index: dict) -> list:
    """One row per served sector, in SERVED order. PURE."""
    out = []
    bench = (body or {}).get("benchmark")
    for s in (body or {}).get("sectors") or []:
        if not isinstance(s, dict):
            continue
        rel_1d = s.get("rel_1d")
        d1 = _num(rel_1d)
        h = _heat_of(s.get("group"), index)
        names = s.get("names") or []
        first = names[0] if names and isinstance(names[0], dict) else None
        out.append({
            "sector": s.get("group"), "n": s.get("n_full"), "benchmark": bench,
            "rel_1d": rel_1d, "rel_5d": s.get("rel_5d"), "rel_21d": s.get("rel_21d"),
            "pct_positive_1d": s.get("pct_positive_1d"),
            "read": {"1d": leg_word(rel_1d), "5d": leg_word(s.get("rel_5d")),
                     "21d": leg_word(s.get("rel_21d"))},
            "heat": h,
            "hot_lagging_1d": h["tone"] == "hot" and d1 is not None and d1 < 0,
            "cold_leading_1d": h["tone"] == "cold" and d1 is not None and d1 > 0,
            "leader": first.get("symbol") if first else None,
            "day_tag": s.get("day_tag"),
        })
    return out


def headlines_block(res: Optional[dict]) -> dict:
    """Trim a `core.search` result to what the tab prints. PURE."""
    if not isinstance(res, dict):
        return {"ok": False, "reason": "headlines unavailable", "items": []}
    items = []
    for it in res.get("items") or []:
        if not isinstance(it, dict):
            continue
        items.append({k: it.get(k) for k in ("title", "url", "source", "published", "provider")})
    return {"ok": True, "items": items, "window_hours": res.get("window_hours"),
            "query": res.get("query"), "counts": res.get("counts"),
            "fetched_at": res.get("fetched_at")}


def _timed_out() -> dict:
    return {"ok": False, "reason": TIMED_OUT.format(LEG_BUDGET_SEC)}


# ── the four legs (each runs in ONE worker thread) ─────────────────────────
def _gauge_leg() -> dict:
    from sepa import market_gauge
    return verdict_block(market_gauge.get_gauge())


def _macro_leg() -> dict:
    return macro_block()


def _sectors_leg() -> dict:
    """Mirrors `rotation/api.py` /rotation/hottest: never builds the member
    table, reads the persisted payload ONCE and reuses it for the heat index,
    and wraps the day-tag attach exactly as the route does."""
    from rotation import api as RA
    from rotation import sector_news_tags as SNT
    table, meta = RA._members_table()
    meta = meta if isinstance(meta, dict) else {}
    if table is None:
        return {"ok": False, "reason": meta.get("reason") or "member table unavailable",
                "rows": []}
    base = RA._members_payload() or {}
    payload = dict(base)
    payload[T.MEMBERS_KEY] = table
    body = H.build_live(payload, sort=H.DEFAULT_SORT, direction=H.DEFAULT_DIR,
                        names_per_group=1, basis=H.D1_CLOSE)
    try:
        SNT.attach(body, SNT.latest_within())
    except Exception as exc:                                   # noqa: BLE001
        log.warning("news_tab: day tags unavailable: %s", exc)
        for s in body.get("sectors") or []:
            if isinstance(s, dict):
                s.setdefault("day_tag", None)
    index = heat.build_index(base)
    rows = sector_rows(body, index)
    tags_date = next((r["day_tag"].get("date") for r in rows
                      if isinstance(r.get("day_tag"), dict) and r["day_tag"].get("date")), None)
    return {"ok": True, "as_of": body.get("as_of"), "benchmark": body.get("benchmark"),
            "ranked_by": H.DEFAULT_SORT, "heat_window": heat.HEAT_WINDOW,
            "d1": d1_block(body), "rows": rows, "tags_date": tags_date,
            "study": SECTOR_STUDY, "source": meta.get("source"),
            "built_at_iso": meta.get("built_at_iso"), "stale": meta.get("stale")}


def _news_leg() -> dict:
    # search_sync is core's own sync wrapper around core.search — the audit
    # write inside it is blocking Mongo, so it runs in this worker thread.
    return headlines_block(core.search_sync(keyword=MARKET_QUERY, audit=AUDIT_TAG))


def _settle(res) -> dict:
    if isinstance(res, asyncio.TimeoutError):
        return _timed_out()
    if isinstance(res, BaseException):
        return {"ok": False, "reason": str(res)[:200] or type(res).__name__}
    return res


async def build() -> dict:
    """The whole tab. Every leg gets `LEG_BUDGET_SEC`; a slow or broken leg
    costs its own block and never the other three."""
    legs = (_gauge_leg, _macro_leg, _sectors_leg, _news_leg)
    got = await asyncio.gather(
        *(asyncio.wait_for(asyncio.to_thread(fn), LEG_BUDGET_SEC) for fn in legs),
        return_exceptions=True)
    verdict, macro, sectors, headlines = (_settle(r) for r in got)
    body = {
        "generated_at_iso": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "verdict": verdict, "macro": macro, "sectors": sectors, "headlines": headlines,
        "budget_sec": LEG_BUDGET_SEC, "note": NOTE, "measured": False,
    }
    # 🧠 the local model's stored two-sided read. A Mongo read on the request;
    # the model itself only ever runs in news_model_read's background thread.
    try:
        read = await asyncio.wait_for(asyncio.to_thread(news_model_read.served, body), LEG_BUDGET_SEC)
    except Exception as exc:                                   # noqa: BLE001
        read = _settle(exc)
    body["model_read"] = read
    return body

File: frontend/src/lib/newsTab.ts
/* newsTab — pure helpers for the 📰 News tab on Chart Maps (2026-09-24).
 *
 * Ajay 2026-09-24: "build me a news tab in chartmaps to give me a bullish
 * market or bearsish market and also pull Macro calendar that has T1 and T2
 * tier events in to this tab consider in to news. If bullish or beaish I need
 * to whcih sectors are bullish or which hotsectors are bearish. In a table."
 *
 * Backend: backend/chart_maps/news_tab.py (GET /chart-maps/news). The SERVER
 * owns every word: the gauge state → market word map lives there, once, and
 * the sector-leg words and the 🔥/🧊 heat tone are served per row. This file
 * only formats what arrived, so a state name typed here could never drift
 * from the one the gauge actually serves — which is why `state` and `word`
 * are plain strings below and never a union of the gauge's states.
 *
 * "today" is printed ONLY when the served day block says the session is live
 * (`d1.live === true`). Before the open, after the close and on weekends the
 * day leg is the LAST CLOSE, and calling it "today" would be a lie on his
 * morning read.
 *
 * Nothing here is measured, nothing gates anything. Not advice.
 */
import type { HsDayTag } from '../components/HottestSectors';

export type NtWord = {
  score?: number | null;
  state?: string | null;
  state_label?: string | null;
  word?: string | null;
};

export type NtVerdict = {
  ok: boolean;
  reason?: string | null;
  as_of_label?: string | null;
  generated_at_iso?: string | null;
  daily?: NtWord | null;
  weekly?: NtWord | null;
  agree?: boolean | null;
  drivers?: string[] | null;
  outlook?: { label?: string | null; note?: string | null; watch?: string[] | null } | null;
  disclaimer?: string | null;
};

export type NtMacroEvent = {
  date: string;
  kind?: string | null;
  tier: number;
  tier_label?: string | null;
  label: string;
  days_until?: number | null;
  when_label?: string | null;
};

export type NtMacro = {
  ok: boolean;
  reason?: string | null;
  days?: number | null;
  tier_labels?: Record<string, string> | null;
  next_tier1?: Partial<NtMacroEvent> | null;
  events?: NtMacroEvent[] | null;
  disclaimer?: string | null;
};

export type NtD1 = {
  live?: unknown;
  basis?: string | null;
  as_of?: string | null;
  reason?: string | null;
  market_closed?: unknown;
};

export type NtHeat = {
  tone?: string | null;
  percentile?: number | null;
  heat_window?: string | null;
  thin?: boolean | null;
  grain?: string | null;
};

/* The benchmark the server SENDS is the rotation grid's own object —
 * `{symbol: 'RSP', window, d1, d5, d21, d63}`, the same shape /rotation serves —
 * not a string. It was typed `string` and rendered straight into JSX, so the
 * first REAL payload threw "Objects are not valid as a React child" and the
 * whole tab rendered blank; every synthetic fixture had used the string 'RSP'.
 * Caught 2026-09-24 by rendering the captured live payload, after 273 unit
 * tests and a review had passed. A bare string is still accepted, so an old
 * payload or a fixture cannot crash the tab either. */
export type NtBenchmark = string | { symbol?: string | null; [k: string]: unknown } | null;

/** The benchmark's ticker, from either shape. Falls back to 'RSP' — the rotation
 *  engine's one fixed benchmark — so a timed-out sectors block still names the
 *  yardstick its columns would have been measured against. */
export function benchmarkSymbol(b: NtBenchmark | undefined): string {
  if (typeof b === 'string') return b.trim() || 'RSP';
  if (b && typeof b === 'object' && typeof b.symbol === 'string' && b.symbol.trim()) {
    return b.symbol.trim();
  }
  return 'RSP';
}

export type NtSectorRow = {
  sector: string;
  n?: number | null;
  benchmark?: NtBenchmark;
  rel_1d?: number | null;
  rel_5d?: number | null;
  rel_21d?: number | null;
  pct_positive_1d?: number | null;
  read?: Record<string, string> | null;
  heat?: NtHeat | null;
  hot_lagging_1d?: boolean | null;
  cold_leading_1d?: boolean | null;
  leader?: string | null;
  day_tag?: HsDayTag | null;
};

export type NtSectors = {
  ok: boolean;
  reason?: string | null;
  as_of?: string | null;
  benchmark?: NtBenchmark;
  ranked_by?: string | null;
  heat_window?: string | null;
  d1?: NtD1 | null;
  rows?: NtSectorRow[] | null;
  tags_date?: string | null;
  study?: ({ note?: string | null } & Record<string, unknown>) | null;
  source?: string | null;
  built_at_iso?: string | null;
  stale?: boolean | null;
};

export type NtHeadline = {
  title?: string | null;
  url?: string | null;
  source?: string | null;
  published?: number | null;
  provider?: string | null;
};

export type NtHeadlines = {
  ok: boolean;
  reason?: string | null;
  items?: NtHeadline[] | null;
  window_hours?: number | null;
  query?: string | null;
  counts?: Record<string, unknown> | null;
  fetched_at?: number | string | null;
};

/** 🧠 the local abliterated model's stored two-sided read
 *  (backend/chart_maps/news_model_read.py). Prose only — the server refused
 *  any read that wrote a number it was not handed, was one-sided, or leaned
 *  anything but bullish / bearish / mixed. UNMEASURED. */
export type NtModelRead = {
  lean?: string | null;
  bull?: string | null;
  bear?: string | null;
  sectors_bullish?: string[] | null;
  sectors_bearish?: string[] | null;
  watch?: string[] | null;
  model?: string | null;
  provider?: string | null;
  generated_at_iso?: string | null;
};

export type NtModelReadBlock = {
  ok: boolean;
  reason?: string | null;
  read?: NtModelRead | null;
  refreshing?: boolean | null;
  age_sec?: number | null;
  last_error?: string | null;
  waiting?: string | null;
  refresh_min_sec?: number | null;
  note?: string | null;
};

export type NewsTabPayload = {
  generated_at_iso?: string | null;
  verdict?: NtVerdict | null;
  macro?: NtMacro | null;
  sectors?: NtSectors | null;
  headlines?: NtHeadlines | null;
  model_read?: NtModelReadBlock | null;
  budget_sec?: number | null;
  note?: string | null;
  measured?: boolean;
};

export type WordTone = 'up' | 'down' | 'flat' | 'muted';

/** Colour for a SERVED word. The words are the server's (market: bullish /
 *  mixed / bearish / unknown; sector leg: bullish / bearish / flat / unknown);
 *  anything else — including "unknown" and an empty string — is muted rather
 *  than guessed at. */
export function wordTone(word: string | null | undefined): WordTone {
  switch ((word || '').trim().toLowerCase()) {
    case 'bullish': return 'up';
    case 'bearish': return 'down';
    case 'mixed':
    case 'flat': return 'flat';
    default: return 'muted';
  }
}

function scoreText(v: number | null | undefined): string {
  return v == null || !Number.isFinite(v) ? '' : String(Math.round(v));
}

/** "mixed · Caution 60" — the served word, then the gauge's OWN label and
 *  score beside it, so the mapping is always visible. */
export function verdictLine(v: NtWord | null | undefined): string {
  const word = (v?.word || '').trim() || 'unknown';
  const tail = [v?.state_label || '', scoreText(v?.score)].filter(Boolean).join(' ');
  return tail ? `${word} · ${tail}` : word;
}

function labelScore(v: NtWord): string {
  return [v.state_label || '', scoreText(v.score)].filter(Boolean).join(' ');
}

/** Whether the daily and the weekly read agree, said in words. Empty when the
 *  weekly read did not arrive — one card, no comparison. Agreement is the
 *  gauge's STATE, compared as served; a missing state never "agrees". */
export function agreeLine(daily: NtWord | null | undefined, weekly: NtWord | null | undefined): string {
  if (!daily || !weekly) return '';
  const dw = (daily.word || '').trim() || 'unknown';
  const ww = (weekly.word || '').trim() || 'unknown';
  const same = daily.state != null && daily.state !== '' && daily.state === weekly.state;
  if (same) {
    return `Daily and weekly agree — ${dw}${daily.state_label ? ` (${daily.state_label})` : ''}`;
  }
  const d = labelScore(daily);
  const w = labelScore(weekly);
  return `Daily and weekly disagree — daily ${dw}${d ? ` (${d})` : ''} · weekly ${ww}${w ? ` (${w})` : ''}`;
}

/** The name of the day leg. "today" ONLY when the server says the session is
 *  live, strictly `true` — a string "yes" or a missing block is the last close. */
export function d1Label(d1: NtD1 | null | undefined): 'today' | 'last close' {
  return d1?.live === true ? 'today' : 'last close';
}

export const SECTOR_VIEWS = ['all', 'hot', 'hot_lagging', 'cold_leading'] as const;
export type SectorView = typeof SECTOR_VIEWS[number];

/** The rows one view chip shows, in SERVED order (nothing is re-sorted). The
 *  two cross-read views read the server's leg-named flags and nothing else. */
export function filterRows(rows: NtSectorRow[] | null | undefined, view: SectorView): NtSectorRow[] {
  const all = rows || [];
  switch (view) {
    case 'hot': return all.filter((r) => r?.heat?.tone === 'hot');
    case 'hot_lagging': return all.filter((r) => r?.hot_lagging_1d === true);
    case 'cold_leading': return all.filter((r) => r?.cold_leading_1d === true);
    default: return all.slice();
  }
}

/** The chip text for one view; the day word follows the served `d1.live`. */
export function viewLabel(view: SectorView, d1: NtD1 | null | undefined): string {
  const day = d1Label(d1);
  switch (view) {
    case 'hot': return '\u{1F525} Hot';
    case 'hot_lagging': return `\u{1F525} hot, lagging ${day}`;
    case 'cold_leading': return `\u{1F9CA} cold, leading ${day}`;
    default: return 'All';
  }
}

/** 🔥 / 🧊 / — for the served tone; "?" for anything the heat engine did not
 *  answer (sector missing from the index, an unknown tone). Never a guess. */
export function heatGlyph(tone: string | null | undefined): string {
  if (tone === 'hot') return '\u{1F525}';
  if (tone === 'cold') return '\u{1F9CA}';
  if (tone === 'neutral') return '—';
  return '?';
}

/** "2026-09-29 · in 5 days" — the served date, then the served when-label. */
export function macroWhen(ev: Partial<NtMacroEvent> | null | undefined): string {
  if (!ev) return '';
  const date = ev.date || '';
  const when = ev.when_label || '';
  return [date, when].filter(Boolean).join(' · ');
}

/** "3h ago" off an epoch (seconds or ms). Empty for a missing or broken stamp. */
export function publishedAgo(ts: number | null | undefined, now: number = Date.now()): string {
  if (ts == null || !Number.isFinite(ts)) return '';
  const ms = ts > 1e12 ? ts : ts * 1000;
  const mins = Math.floor((now - ms) / 60000);
  if (mins < 1) return 'just now';
  if (mins < 60) return `${mins}m ago`;
  const hours = Math.floor(mins / 60);
  if (hours < 48) return `${hours}h ago`;
  return `${Math.floor(hours / 24)}d ago`;
}

/** "12 min ago" off the SERVED age in seconds; '' for anything unusable. */
export function ageLabel(sec: number | null | undefined): string {
  if (sec == null || !Number.isFinite(sec) || sec < 0) return '';
  if (sec < 60) return 'just now';
  if (sec < 3600) return `${Math.floor(sec / 60)} min ago`;
  if (sec < 86400) return `${Math.floor(sec / 3600)} h ago`;
  return `${Math.floor(sec / 86400)} d ago`;
}

/** Served strings only, de-duplicated; junk entries dropped. */
export function nameList(v: unknown): string[] {
  if (!Array.isArray(v)) return [];
  const out: string[] = [];
  for (const x of v) if (typeof x === 'string' && x.trim() && !out.includes(x.trim())) out.push(x.trim());
  return out;
}

export const MODEL_WRITING =
  'the local model is writing a fresh read — about 2–3 min on this model, refresh then';
